pkg/agent/agentclient/client.go

```diff
@@ -85,31 +92,34 @@ func (a *AgentClient) Serve(stopCh <-chan struct{}) {
 
 			connID := atomic.AddInt64(&a.nextConnID, 1)
 			a.conns[connID] = conn
+			a.dataChs[connID] = make(chan []byte, 5)
+			var once sync.Once
+
+			cleanup := func() {
+				once.Do(func() {
+					conn.Close()
+					close(a.dataChs[connID])
+					delete(a.conns, connID)
+					delete(a.dataChs, connID)
+				})
+			}
 
 			resp.GetDialResponse().ConnectID = connID
 			if err := a.stream.Send(resp); err != nil {
 				glog.Warningf("stream send error: %v", err)
+				continue
```

Comment:
Could this become an infinite loop on a broken stream?

Reply:
no, it will go back to `a.stream.Recv()` and exit on error there.